python-squid/src/test/java/org/sonar/plugins/python/api/cfg/ControlFlowGraphTest.java

```diff
@@ -404,6 +405,168 @@ public void raise_in_try() {
+      "  except_block(succ = [finally_block], syntSucc = finally_block)",
+      "  return",
+      "finally:",
+      "  finally_block(succ = [after_try, END])",
+      "  pass",
+      "after_try(succ = [END])");
+  }
+
+  @Test
+  public void return_in_else() {
+    verifyCfg(
+      "before_try(succ = [try_body])",
+      "try:",
+      "  try_body(succ = [else_block, except_cond])",
+      "except except_cond(succ = [except_block, finally_block]):",
+      "  except_block(succ = [finally_block])",
+      "else:",
+      "  else_block(succ = [finally_block], syntSucc = finally_block)",
+      "  return",
+      "finally:",
+      "  finally_block(succ = [after_try, END])",
+      "  pass",
+      "after_try(succ = [END])");
+  }
+
+
+  @Test
+  public void continue_in_except() {
+    verifyCfg(
+      "before_while(succ = [cond])",
+      "while cond(succ = [try_body, after_while]):",
+      "  try:",
+      "    try_body(succ = [finally_block, except_cond])",
+      "  except except_cond(succ = [except_block, finally_block]):",
+      "    except_block(succ = [finally_block], syntSucc=finally_block)",
+      "    continue",
+      "  finally:",
+      "    finally_block(succ = [after_try, END])",
+      "  after_try(succ = [cond])",
+      "after_while(succ = [END])");
+  }
+
+  @Test
+  public void continue_in_else() {
+    verifyCfg(
+      "before_while(succ = [cond])",
+      "while cond(succ = [try_body, after_while]):",
+      "  try:",
+      "    try_body(succ = [else_block, except_cond])",
+      "  except except_cond(succ = [except_block, finally_block]):",
+      "    except_block(succ = [finally_block])",
+      "  else:",
+      "    else_block(succ = [finally_block], syntSucc=finally_block)",
+      "    continue",
+      "  finally:",
+      "    finally_block(succ = [after_try, END])",
+      "  after_try(succ = [cond])",
+      "after_while(succ = [END])");
+  }
+
+  @Test
+  public void break_in_except() {
+    verifyCfg(
+      "before_while(succ = [cond])",
+      "while cond(succ = [try_body, after_while]):",
+      "  try:",
+      "    try_body(succ = [finally_block, except_cond])",
+      "  except except_cond(succ = [except_block, finally_block]):",
+      "    except_block(succ = [finally_block], syntSucc=finally_block)",
+      "    break",
+      "  finally:",
+      "    finally_block(succ = [after_try, END])",
+      "  after_try(succ = [cond])",
+      "after_while(succ = [END])");
+  }
+
+  @Test
+  public void break_in_else() {
+    verifyCfg(
+      "before_while(succ = [cond])",
+      "while cond(succ = [try_body, after_while]):",
+      "  try:",
+      "    try_body(succ = [else_block, except_cond])",
+      "  except except_cond(succ = [except_block, finally_block]):",
+      "    except_block(succ = [finally_block])",
+      "  else:",
+      "    else_block(succ = [finally_block], syntSucc=finally_block)",
+      "    break",
+      "  finally:",
+      "    finally_block(succ = [after_try, END])",
```

Comment:
I believe the successors are wrong here because of the `break`. I think it should be: `[after_while, END]`. And so I believe that `after_try` is dead code.

Reply:
As discussed, we may have a path: `try -> except_cond -> except_block -> finally_block -> after_try`.  Successors of `finally_block` are indeed not precise enough, this is a limitation of our current implementation. I will put a comment about it.